---
title: "Introduction to Transcriptomics: Part I"
author: Nicole Black, Wade Boohar
date: 07/24/22
output: pdf_document
updated: 02/26/23
---

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/Users/soniaqi/Documents/QBio-490/qbio_490_sonia/analysis_data")) 
```

```{r Package Setup}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
library(SummarizedExperiment)
library(TCGAbiolinks)
```

Load in the RNAseq data

```{r}
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

#GDCdownload(rna_query)

rna_se <- GDCprepare(rna_query)

```


###### (1) Summarized Experiment Basics

Let's take a look at rna_se (RNA Summarized Experiment). Call head() and str(). Also explore the data frame by double clicking on it from the Environment panel (or calling "View(rna_se)").

```{r Exercise 1.1}
head(rna_se)
str(rna_se)
```

Use the dim() function to find the dimensions of the colData, rowRanges, and unstranded dataframes. Record them below:

  ~ colData (rows, cols): 1231, 87
  ~ rowRanges: 60660, 10
  ~ unstranded: 60660, 1231

```{r Exercise 1.2}
dim(rna_se@colData)
dim(rna_se@rowRanges@elementMetadata)
dim(rna_se@assays@data$unstranded)
```

Run the following code and fill out a description of the rows/columns of our data frames.You can use this as a cheat sheet moving forward.
  ~ colData
        - Rows: patients
        - Columns: clinical information
  ~ rowRanges
        - Rows: different genes
        - Columns: information about the genes
  ~ unstranded (in assays)
        - Rows: different genes
        - Columns: patients
  
```{r Exercise 1.3}
# colData: 
rownames(rna_se@colData)
colnames(rna_se@colData)

# rowRanges:
str(rna_se@rowRanges@elementMetadata)
# what are the rows... use your logic
rownames(rna_se@assays@data$unstranded)
colnames(rna_se@assays@data$unstranded)
# for unstranded, don't call anything, just use your logic from exercise 1.2 and and think it through!
```

###### (2) ColData

View the first 15 values in the age_at_index column.

```{r Exercise 2.1}
rna_se@colData[1:15, 'age_at_index']
```

Call summary on the age_at_index column. Record the output below:
    Min.: 26.00
    1st Qu.: 48.00
    Median: 58.00
    Mean: 58.27
    3rd Qu.: 67.00
    Max.: 90.00
    NA's: 1
    
```{r Exercise 2.2}
summary(rna_se@colData[,'age_at_index'])
```

Create a data frame called rna_clinical that contains data from colData, except for the NA patient and their data.

```{r Exercise 2.3}
rna_clinical <- rna_se@colData
```

Use the as.data.frame() function to change this from an S4 object to a data frame object.

```{r Exercise 2.4}
rna_clinical <- as.data.frame(rna_clinical)
```

Make a mask and subset out the treatments column.

```{r Exercise 2.5}
treatments_mask <- colnames(rna_clinical) %in% c('treatments', 'primary_site', 'disease_type')
rna_clinical <- rna_clinical[, !treatments_mask]
```

Save rna_clinical to your local machine

```{r Exercise 2.6}
write.csv(rna_clinical, "brca_rna_clincial_data.csv", row.names = TRUE)
```

###### (3) RowRanges

Call head() on rowRanges@elementMetadata:

```{r Exercise 3.1}
head(rna_se@rowRanges@elementMetadata)
```

Create a data frame called rna_genes that contains data from elementMetadata.

```{r Exercise 3.2}
rna_genes <- rna_se@rowRanges@elementMetadata
```

Use the as.data.frame() function to change this from an S4 object to a data frame object

```{r Exercise 3.3}
rna_genes <- as.data.frame(rna_genes)
```

Save rna_genes to your local machine.

```{r Exercise 3.4}
write.csv(rna_genes, "brca_rna_gene_data.csv", row.names = FALSE)
```

###### (4) Assays

What are the names of the 6 assays?
unstranded, stranded_first, stranded_second, tpm_unstrand, fpkm_unstrand, fpkm_uq_unstrand
```{r Exercise 4.1}
rna_se@assays@data$
```

Create a new data frame called "rna_counts" that contains the data from unstranded. Recall the patient with NA age data? We need to exclude that patient. We can use the same mask to do this.

```{r Exercise 4.2}
rna_counts <- rna_se@assays@data$unstranded
```

Fetch the data for the first 5 patients and the first 15 rows.

```{r Exercise 4.3}
rna_counts[1:15, 1:5]
```

Save rna_counts to your local machine (warning: this will likely take a while!)

```{r Exercise 4.4}
write.csv(rna_counts, "brca_rna_count_data.csv", row.names = TRUE)
```


###### (5) Cleaning up your data frames

Call rownames() and colnames() on rna_clinical.

```{r Exercise 5.1}
rownames(rna_clinical)
colnames(rna_clinical)
```

As you can see, the column names are informative, but the row names are not. Let's fix this. Use the row.names() function to name the rows according to the barcode column.

```{r Exercise 5.2}
?row.names

# the row names are already descriptive
```

Just like we did to "clinical", we are going to create a column in rna_clinical for age_category. Our cutoff will be the same ("young" as <= 58 y/o, and "old" as > 58 y/o).

There are three age columns in rna_clinical. Which one should we use? Explore the rna_clinical data frame and find the three age related columns. Call summary() on all of them and based on the results, determine which column to use.

```{r Exercise 5.3}
rna_clinical$age_category <- ifelse(rna_clinical$age_at_index>58, 'Old', 'Young')

# Hint: if you're really stuck finding the columns/determining which to use, go to the bottom of the "Cleaning up your data frames" section for the answers. Only do this once you've given it a good shot on your own!
```

Now that you have your existing data column. Use this with an ifelse() statement to create a new column in rna_clinical called age_category. Remember ifelse(expression, x, y), where x happens if the expression is true, and y happens if it is false.

```{r Exercise 5.4}
rna_clinical$age_category
```

----

Let's move on to rna_genes. Call rownames() and colnames() on rna_genes.

```{r Exercise 5.5}
rownames(rna_genes)
colnames(rna_genes)
```

Once again, the column names are informative, but the row names are not. Use row.names() to name the rows according to gene_id. Note: it would be nice to use gene_name instead of gene_id for the row names, but there are duplicate values in the gene_name column, and row names must be unique.

```{r Exercise 5.6}
row.names(rna_genes) <- rna_genes$gene_id
```

----

Now that rna_genes has informative row/column names, let's move on to rna_counts. Call rownames() and colnames() on rna_counts.

```{r Exercise 5.7}
rownames(rna_counts)
colnames(rna_counts)
```

Uh oh! Looks like neither the row nor column names are very informative. Given what we did to the other data frames, add column and row names to rna_counts. Name the columns based on the patient barcodes (ie the column names for rna_clinical), and name the rows based on the gene id's (ie the column names for rna_gene).

```{r Exercise 5.8}
colnames(rna_counts) <- rna_clinical$barcode
rownames(rna_counts) <- rna_genes$gene_id
```

----

The last step in cleanup is to subset out any samples that are of a normal tissue type. Call unique() on the definition column of rna_clinical.

```{r Exercise 5.9}
unique(rna_clinical$definition)
```

Notice that there are three sample types: "Solid Tissue Normal", "Primary solid Tumor", and "Metastatic". For now, we are simply going to get rid of the "Solid Tissue Normal" samples, but keep this in mind for your midterm and final projects as these samples could provide you with a good control!

Create a mask that calls "Solid Tissue Normal" FALSE, and the other two sample types TRUE.

```{r Exercise 5.10}
tissue_mask <- ifelse(rna_clinical$definition == 'Solid Tissue Normal',F,T)
rna_clinical <- rna_clinical[tissue_mask,]
```

You will need to apply this mask to two of our three data frames (rna_clinical, rna_genes, rna_counts). Think about which df's involve samples, and apply the mask to these data frames to subset out the normal samples. You may overwrite the data frames in this case.

```{r Exercise 5.11}
rna_counts <- rna_counts[, tissue_mask]
```

Great! Now all of our data frames are ready for exploration and analysis.

-- Hints --

Age columns:
  rna_clinical$age_at_diagnosis (16 NAs) (in days, not years)
  rna_clinical$age_at_index (0 NAs) <-- use this!
  rna_clinical$paper_age_at_initial_pathologic_diagnosis (132 NAs)

###### (6) Preparing for plotting

Make a column in rna_clinical called five_yr_surv that contains TRUEs and FALSEs for whether a patient survived 5 years. Use days_to_death, or (if days_to_death is NULL), days_to_last_follow_up. Note: 1 year = 365.25 days.

```{r Exercise 6.1}
# fill in the nested ifelse to capture TRUE/FALSE values for whether a patient survived 5 years according to their days_to_death (NOTE: the first ifelse condition changes the NAs from next to true NA values so we can use is.na() later on)

five_yr_death <- ifelse(rna_clinical$days_to_death == "NA", NA, ifelse(rna_clinical$days_to_death>5*365.25,T,F))

# use boolean data from days_to_last_followup when five_yr_death is NA

five_yr_death_and_followup <- ifelse(is.na(five_yr_death), ifelse(rna_clinical$days_to_last_follow_up>365.25*5,T,F), five_yr_death)

# create your column from your finalized boolean vector (there should be only T/F, no NAs) 
rna_clinical$five_year_surv <- five_yr_death_and_followup
```

----

Recall the %in% function that allows us to check whether a value is present in a vector. Ex:

```{r Exercise 6.2}
fruit_list <- c("peach", "pomegranate", "cherry", "apricot", "kiwi")
"apple" %in% fruit_list # output: FALSE
"kiwi" %in% fruit_list # output: TRUE
```

Using your knowledge from the literature presentations, Google, etc., pick two genes that are important in breast cancer. List them below.
    ~ Gene 1: TP53
    ~ Gene 2: MUC16
    
Using the %in% function, check that these two genes are in the rna_genes data. If they aren't pick a new gene. Hint: think about which column to check in.

```{r Exercise 6.3}
c('TP53','MUC16') %in% rna_genes$gene_name
```

Create a mask for geneA and geneB where that gene is TRUE and all others are FALSE.

NOTE: for the rest of this notebook, whenever you see "geneA" or "geneB", change that to the name of your gene(s) for easier reference later on.

```{r Exercise 6.4}
# write code here

TP53_mask <- rna_genes$gene_name == 'TP53'
# create mask based on the gene name (your mask should be true when the name is 'geneA' and false for any other gene name)

MUC16_mask <- rna_genes$gene_name == 'MUC16'
```

We can use this mask for a few things. First, let's use it to grab the ensembl IDs of our genes.

```{r Exercise 6.5}
TP53_ensembl <- rna_genes$gene_id[TP53_mask] # think about which column we should apply out mask to
  
MUC16_ensembl <- rna_genes$gene_id[MUC16_mask]
```

We can also use the mask to create a list that contains counts for just a single gene. (HINT: this means subsetting for just the geneA row, with all columns included)

```{r Exercise 6.6}
TP53_counts <- rna_counts[TP53_ensembl,]

MUC16_counts <- rna_counts[MUC16_ensembl,]
```

Call summary() on geneA_counts and geneB_counts. Record the output below:

  Gene A
    ~ Min.: 274
    ~ 1st Qu.: 2878
    ~ Median: 4444
    ~ Mean: 5002
    ~ 3rd Qu.: 6495
    ~ Max.: 25897
    
  Gene B
  ~ Min.: 0
  ~ 1st Qu.: 35.25
  ~ Median: 143
  ~ Mean: 1986.54
  ~ 3rd Qu.: 880.75
  ~ Max.: 73185

```{r Exercise 6.7}
summary(TP53_counts)
summary(MUC16_counts)
```

###### (7) Plotting

Recall that plot() creates a scatterplot. Use this function to plot geneA vs geneB.

```{r Exercise 7.1}
# par(mar=c(1,1,1,1)) # this changes your margins so the plot will appear in the window correctly
plot(TP53_counts,
     MUC16_counts,
     xlab = "TP53 Counts",
     ylab = "MUC16 Counts",
     main = "TP53 vs MUC16"
)
```

Do you notice any patterns? Is there a relationship between the two genes that you chose? Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:
Most data points are concentrated in the bottom left corner. Specifically, most TP53 counts and MUC16 counts are below 10,000. 
----

Let's visualize these two genes again, this time using box plots. To get both geneA_counts and geneB_counts on one plot, make a data frame using data.frame() that contains those lists as columns.

```{r Exercise 7.2}
geneAB_counts <- data.frame(TP53_counts,MUC16_counts) # create data frame with geneA_counts and geneB_counts

# par(mar=c(1,1,1,1))
boxplot(geneAB_counts,
        xlab = "Gene",
        ylab = "Counts")
```

Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:
MUC16 counts has a lower center and a smaller IQR range than TP53 counts. Also, MUC16 counts has a lot more larger outliers than that of TP53. 
----

For the next plots, choose either gene A or gene B to continue with. We will be making a box plot with two boxes, one for patients who survived 5 years, and another for patients who did not.

```{r Exercise 7.3}
# par(mar=c(1,1,1,1))
boxplot(MUC16_counts ~ rna_clinical$five_year_surv, # think about what data (gene counts or the 5 yr survival column) should be on the x vs y axes
        xlab = "Five Year Survival",
        ylab = "MUC16 Counts")
# dev.off()
```

Does there appear to be a relationship between expression of your gene and five year survival? Is this what we would expect (do some research on the gene+survival to validate). Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:
The centers for the two distributions are very similar, although the IQR of the survived group seems to be slightly larger. The group that did not live 5 years has distribution more skewed to the right (larger MUC16 counts) with more outliers. This means that a large number of MUC16 mutation counts may decrease the survival chance of patients. 
---- 

Create two more gene_count lists, for MKI67 and PTGS2.

```{r Exercise 7.4}
MKI67_mask <- rna_genes$gene_name == 'MKI67'
PTGS2_mask <- rna_genes$gene_name == 'PTGS2'

MKI67_ensembl <- rna_genes$gene_id[MKI67_mask]
PTGS2_ensembl <- rna_genes$gene_id[PTGS2_mask]
MKI67_counts <- rna_counts[MKI67_ensembl,]
PTGS2_counts <- rna_counts[PTGS2_ensembl,]
```

Create a Draftsman plot by filling in the code below.

```{r Exercise 7.5}
# fill in with your 4 individual gene count lists and their names
geneABCD_counts <- data.frame(TP53_counts,MUC16_counts,MKI67_counts,PTGS2_counts)
colnames(geneABCD_counts) <- c('TP53','MUC16','MKI67','PTGS2')

# the following three lines set up a clustering color scheme for our plot
# choose a column from rna_clinical to cluster based on (ex: age_category, five_yr_surv, AJCC staging, etc.)
cols <- character(nrow(rna_clinical)) 
cols[rna_clinical$gender == 'female'] <- "blue" # fill in column and value for as many unique values are in your column (ex: 2 for age_category, "young"/"old")
cols[rna_clinical$gender == 'male'] <- "red"

# run this line to create your plot!
cols[228] <- NA
pairs(geneABCD_counts, col = cols, lower.panel=NULL)
```

Do there appear to be any relationships between the genes in question? Do you see any clear clustering based on your chosen clinical variables? Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:
There seems to be a negative association between counts for TP53 and MUC16 and MUC16 and MKI67 (probably a less r^2 value for this pair). There also seems to be a positive association between TP53 and MKI67. There does not seem to be any obvious association between counts of PTGS2 and other genes since the pattern is quite vertical.  

###### (8) Saving your edited files

Write your df files to your local computer (you may overwrite the existing files). Use write.csv().

```{r Exercise 8.1}
write.csv(rna_clinical,"/Users/soniaqi/Documents/QBio-490/qbio_490_sonia/analysis_data/brca_rna_clinical_data.csv", row.names = FALSE)
write.csv(rna_counts,"/Users/soniaqi/Documents/QBio-490/qbio_490_sonia/analysis_data/brca_rna_count_data.csv", row.names = FALSE)
write.csv(rna_genes,"/Users/soniaqi/Documents/QBio-490/qbio_490_sonia/analysis_data/brca_rna_gene_data.csv", row.names = FALSE)
```
